### PR TITLE
[go_router] Adds `reverseTransitionDuration` to `CustomTransitionPage`

### PR DESCRIPTION
**Pull request by heeyunlee**
This PR adds `reverseTransitionDuration` to `CustomTransitionPage` so that you can set different duration for a `transitionDuration` and `reverseTransitionDuration` and enables more customization using `CustomTransitionPage`. It allows us to do something like below:

https://user-images.githubusercontent.com/32585133/217086560-2852d5ba-3c61-48fc-9453-2a537f5eab0e.mov

Closes https://github.com/flutter/flutter/issues/118212

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/test/custom_transition_page_test.dart [hunk 1]
```diff
@@ -117,6 +117,45 @@ void main() {
...
+    expect(find.byKey(homeKey), findsOneWidget);
+
+    router.push('/login');
+    await tester.pumpAndSettle();
+    expect(find.byKey(loginKey), findsOneWidget);
+
+    router.pop();
+    await tester.pumpAndSettle();
+    expect(find.byKey(homeKey), findsOneWidget);
+  });
```
Comment (heeyunlee) [inline packages/go_router/test/custom_transition_page_test.dart:158]: Having trouble coming up with a test that asserts `transitionDuration` and `reverseTransitionDuration` is different... Any suggestions?

Comment (chunhtai) [reply packages/go_router/test/custom_transition_page_test.dart:158]: Just throwing out idea. In you transition builder, can you build a text widget with animation value, and in the test you can use tester.pump(Duration(milliseconds: 100)) and make sure the animation value is close to reverseDuration/100ms

Comment (heeyunlee) [reply packages/go_router/test/custom_transition_page_test.dart:158]: Thanks for the suggestions! I ended up finding a simpler approach:
``` dart
router.push('/login');
final int pushingPumped = await tester.pumpAndSettle();
expect(find.byKey(loginKey), findsOneWidget);

router.pop();
final int poppingPumped = await tester.pumpAndSettle();
expect(find.byKey(homeKey), findsOneWidget);

expect(pushingPumped != poppingPumped, true);
```

**Review by chunhtai — APPROVED** [review]
LGTM

**Review by hangyujin — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
can you fix the merge conflict?

**Pull request merged**